tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,203 @@
+      get<Tags::deriv<ScalarWave::Phi<spatial_dim>, tmpl::size_t<spatial_dim>,
+                      Frame::Inertial>>(local_derivs);
+
+  // (1.) Get the constraints, and check that they vanish to error_tolerance
+  auto one_index_constraint =
+      make_with_value<tnsr::i<DataVector, spatial_dim, Frame::Inertial>>(
+          x, std::numeric_limits<double>::signaling_NaN());
+  ScalarWave::one_index_constraint(make_not_null(&one_index_constraint),
+                                   deriv_psi, phi);
+  auto two_index_constraint =
+      make_with_value<tnsr::ij<DataVector, spatial_dim, Frame::Inertial>>(
+          x, std::numeric_limits<double>::signaling_NaN());
+  ScalarWave::two_index_constraint(make_not_null(&two_index_constraint),
+                                   deriv_phi);
+
+  Approx numerical_approx = Approx::custom().epsilon(error_tolerance).scale(1.);
+  CHECK_ITERABLE_CUSTOM_APPROX(
+      one_index_constraint,
+      (make_with_value<tnsr::i<DataVector, spatial_dim, Frame::Inertial>>(x,
+                                                                          0.)),
+      numerical_approx);
+  CHECK_ITERABLE_CUSTOM_APPROX(
+      two_index_constraint,
+      (make_with_value<tnsr::ij<DataVector, spatial_dim, Frame::Inertial>>(x,
+                                                                           0.)),
+      numerical_approx);
+
+  // (2.) Test that compute tags return expected values
+  const auto box = db::create<
+      db::AddSimpleTags<
+          ::Tags::Coordinates<spatial_dim, Frame::Inertial>, ScalarWave::Psi,
+          ScalarWave::Phi<spatial_dim>, ScalarWave::Pi,
+          ::Tags::deriv<ScalarWave::Psi, tmpl::size_t<spatial_dim>,
+                        Frame::Inertial>,
+          ::Tags::deriv<ScalarWave::Phi<spatial_dim>, tmpl::size_t<spatial_dim>,
+                        Frame::Inertial>>,
+      db::AddComputeTags<
+          ScalarWave::Tags::OneIndexConstraintCompute<spatial_dim>,
+          ScalarWave::Tags::TwoIndexConstraintCompute<spatial_dim>>>(
+      x, psi, phi, pi, deriv_psi, deriv_phi);
+
+  // Check that their compute items in databox furnish identical values
+  CHECK(db::get<ScalarWave::Tags::OneIndexConstraint<spatial_dim>>(box) ==
+        one_index_constraint);
+  CHECK(db::get<ScalarWave::Tags::TwoIndexConstraint<spatial_dim>>(box) ==
+        two_index_constraint);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.Constraints",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/ScalarWave/"};
+  INFO("Testing constraints with randomized input");
+  // Test the one-index constraint with random numbers
+  const auto used_for_size =
+      DataVector(4, std::numeric_limits<double>::signaling_NaN());
+  test_one_index_constraint_random<1>(used_for_size);
+  test_one_index_constraint_random<2>(used_for_size);
+  test_one_index_constraint_random<3>(used_for_size);
```

Comment:
Just FYI, we have a macro for this:

```suggestion
  GENERATE_UNINITIALIZED_DATAVECTOR;
  CHECK_FOR_DATAVECTORS(test_one_index_constraint_random, (1, 2, 3))
```

(It's in `CheckWithRandomValues.hpp`)

Reply:
thanks for the suggestion :+1: